### PR TITLE
ipfs-http-client: add package

### PR DESCRIPTION
**Pull request by LLE8**
Maintainer: me
Compile and Run tested: TL-MR3020 v3 (mt76x8), OpenWrt current. 

Description:
There are two packages: ipfs-http-client and nlohmannjson.
Package ipfs-http-client is IPFS C++ HTTP API client library, which allows C++ applications to communicate with an IPFS node. It implements IPFS API bindings for C++. Not all methods are implemented. Homepage https://github.com/vasild/cpp-ipfs-http-client
Package nlohmannjson is JSON library for modern C++, yet another JSON library, C++11 header-only. Design goals: intuitive syntax, trivial integration, serious testing. 
Homepage https://github.com/nlohmann/json
API documentation https://nlohmann.github.io/json/
ipfs-http-client depends on this library.

### CODE REVIEW

**Comment by neheb** [comment]
library not needed?

**Comment by karlp** [comment]
Headers only "library"  LLE8, you need a maintainer though, and there's more whhitespace than you really need.  You can probably mark it as PKGARCH:=all too

**Comment by LLE8** [comment]
No *.so or *.a, only c++ headers, like debian package nlohmann-json3-dev. Note, what debian package nlohmann-json3 (without -dev) does not exist.

**Comment by LLE8** [comment]
May be not PKGARCH:=all, but BUILDONLY:=1 ?

**Comment by neheb** [comment]
BUILDONLY sounds right.

**Comment by LLE8** [comment]
Does the variables definitions order matter? What to read?
In developer guide  https://openwrt.org/docs/guide-developer/packages sequence as follows
PKG_NAME, PKG_VERSION, PKG_RELEASE
next
PKG_LICENSE, PKG_LICENSE_FILES
next
PKG_BUILD_DIR, PKG_SOURCE, PKG_SOURCE_URL, PKG_HASH

**Comment by karlp** [comment]
> BUILDONLY sounds right.

where on earth is buildonly documented? I've never heard of this, and it's not on https://openwrt.org/docs/guide-developer/packages

**Comment by LLE8** [comment]
https://archive.openwrt.org/kamikaze/8.09.2/docs/openwrt.html#x1-460002.1.2
May be BUILDONLY parameter is deprecated?
But it still exists in current source code
```
user1@black2:~/openwrt_current$ grep BUILDONLY include/*.mk
include/package-defaults.mk:  BUILDONLY:=
include/package-dumpinfo.mk:)$(if $(BUILDONLY),Build-Only: $(BUILDONLY)
```

**Comment by neheb** [comment]
I just tested it. It works as intended.

File: libs/ipfs-http-client/Makefile [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	$(call Package/$(PKG_NAME)/Default/description)
+	This package contains shared library.
+endef
+
+define Package/$(PKG_NAME)-tests
+	SECTION:=utils
+	CATEGORY:=Utilities
+	TITLE:=IPFS client library tests
+	URL:=https://github.com/vasild/cpp-ipfs-http-client
+	DEPENDS:=+libipfs-http-client +libcurl
```
Comment (karlp) [inline libs/ipfs-http-client/Makefile]: are these tests actualyl something someone would ever want on their target device?

File: libs/ipfs-http-client/patches/001-add-srv-addr.patch [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
++    srvaddr=argv[1];
++
+   try {
+-    ipfs::Client client("localhost", 5001);
++    ipfs::Client client(srvaddr, 5001);
+ 
+     /** [ipfs::Client::FilesGet] */
+     std::stringstream contents;
+--- a/test/generic.cc	2019-11-22 18:16:53.973272202 +0300
++++ b/test/generic.cc	2019-11-22 18:16:58.245249328 +0300
```
Comment (karlp) [inline libs/ipfs-http-client/patches/001-add-srv-addr.patch]: do you really need to patch all the tests? I guess it's the same as "are these tests useful for the openwrt build environment"

File: libs/ipfs-http-client/src/test/demo.cpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  char addr127001[]="127.0.0.1";
+  char *addr;
+  if (argc<2)
+    addr=addr127001;
+  else
+    addr=argv[1];
+  ipfs::Client client(addr, 5001);
+  client.FilesGet("/ipfs/QmYwAPJzv5CZsnA625s3Xf2nemtYgPpHdWEz79ojWnPbdG/readme", &contents);
+  std::cout << contents.str() << std::endl;
+  return 0;
```
Comment (karlp) [inline libs/ipfs-http-client/src/test/demo.cpp]: why would someone want this built and installed?

**Comment by karlp** [comment]
@neheb what "works as intended" buildonly?  it doesn't seem to have any of having any impact, are you sure it does "works as intended" with just the INstallDev section alone?  I mean, sure it puts a field in the opkg metadata, but nothing looks at it.  (And it's still undocumented, so do we really want to encourage it's use?)

**Comment by LLE8** [comment]
ipfs

**Comment by LLE8** [comment]
@karlp
BUILDONLY (optional)
Set this option to 1 if you do NOT want your package to appear in menuconfig. This is useful for packages which are only used as build dependencies.

It is not critical, i can remove it, let it appear in menuconfig. But nlohmannjson does not form a ipkg-package itself.

File: libs/ipfs-http-client/Makefile (hunk 1 above)
Comment (LLE8) [reply libs/ipfs-http-client/Makefile]: It's optional

**Comment by neheb** [comment]
> @neheb what "works as intended" buildonly? it doesn't seem to have any of having any impact, are you sure it does "works as intended" with just the INstallDev section alone? I mean, sure it puts a field in the opkg metadata, but nothing looks at it. (And it's still undocumented, so do we really want to encourage it's use?)

BUILDONLY gets rid of the warning that the package has no Install section. It also gets rid of it from menuconfig.

I have no issue with it.

File: libs/ipfs-http-client/Makefile [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+PKG_NAME:=ipfs-http-client
+PKG_RELEASE:=1
+
+PKG_MAINTAINER:=Leonid Esman <leonid.esman@gmail.com>
+PKG_LICENSE:=MIT
+PKG_LICENSE_FILES:=LICENSE.MIT
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/vasild/cpp-ipfs-http-client.git
+#PKG_SOURCE_DATE:=2019-11-05
```
Comment (neheb) [inline libs/ipfs-http-client/Makefile]: This should be uncommented, no?

Comment (LLE8) [reply libs/ipfs-http-client/Makefile]: I am not sure, this only affects the package name, but not the program functionality.

File: libs/ipfs-http-client/patches/001-add-srv-addr.patch (hunk 2 above)
Comment (LLE8) [reply libs/ipfs-http-client/patches/001-add-srv-addr.patch]: these tests are useless on target device without this patch.

File: libs/ipfs-http-client/src/test/demo.cpp (hunk 3 above)
Comment (LLE8) [reply libs/ipfs-http-client/src/test/demo.cpp]: Because ipfs is experimental software.

**Comment by karlp** [comment]
> @karlp
> BUILDONLY (optional)
> Set this option to 1 if you do NOT want your package to appear in menuconfig. This is useful for packages which are only used as build dependencies.
> 

Where did that documentation come from? It's not in the "creating packages" wiki page, and I can't find it anywhere else

**Comment by LLE8** [comment]
From Kamikaze 8.09
https://archive.openwrt.org/kamikaze/8.09.2/docs/openwrt.html#x1-460002.1.2

File: libs/ipfs-http-client/Makefile (hunk 4 above)
Comment (LLE8) [reply libs/ipfs-http-client/Makefile]: Many packages use PKG_SOURCE_DATE. Let it be.

File: libs/ipfs-http-client/Makefile [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	This is Vasil Dimov's C++ IPFS HTTP API client library.
+endef
+
+define Package/lib$(PKG_NAME)
+	SECTION:=libs
+	CATEGORY:=Libraries
+	TITLE:=IPFS client library
+	URL:=https://github.com/vasild/cpp-ipfs-http-client
+	DEPENDS:= +libcurl +libstdcpp
+	PKG_BUILD_DEPENDS:=nlohmannjson
```
Comment (neheb) [inline libs/ipfs-http-client/Makefile]: I...don't think this belongs here.

File: libs/ipfs-http-client/Makefile [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/vasild/cpp-ipfs-http-client.git
+PKG_SOURCE_DATE:=2019-11-05
+PKG_SOURCE_VERSION:=763e59ad698f3e3846f85df11e01c18ef3fbc401
+PKG_MIRROR_HASH:=a2b5721faf0d43ddb4a892245ef382666149c83f3f97e558e1a6acf2402fb9fd
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/$(PKG_NAME)/Default/description
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
define Package/ipfs-http-client/Default/description
```

See for more details https://github.com/openwrt/packages/pull/4119

File: libs/ipfs-http-client/Makefile [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	IPFS (the InterPlanetary File System) is the Distributed Web. 
+	Specs, docs, sources, links: https://ipfs.io/ and https://github.com/ipfs.
+	This is Vasil Dimov's C++ IPFS HTTP API client library.
+endef
+
+define Package/lib$(PKG_NAME)
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
define Package/libipfs-http-client
```

File: libs/nlohmannjson/Makefile [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	Niels Lohmann's JSON library for modern C++
+endef
+
+define Build/InstallDev
+	$(INSTALL_DIR) $(1)/usr/include
+	$(CP) $(PKG_BUILD_DIR)/include/nlohmann $(1)/usr/include
+endef
+
+$(eval $(call BuildPackage,nlohmannjson))
+
```
Comment (BKPepe) [inline libs/nlohmannjson/Makefile]: Why empty row here? Remove it.

File: libs/nlohmannjson/Makefile [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+define Package/nlohmannjson
+	SECTION:=libs
+	CATEGORY:=Libraries
+	TITLE:=JSON for Modern C++
+	URL:=https://nlohmann.github.io/json/
+	BUILDONLY:=1
+endef
+
+define Package/nlohmannjson/description
+	Niels Lohmann's JSON library for modern C++
```
Comment (BKPepe) [inline libs/nlohmannjson/Makefile]:
```suggestion
  Niels Lohmann's JSON library for modern C++
```

File: libs/nlohmannjson/Makefile [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+PKG_HASH:=e109cd4a9d1d463a62f0a81d7c6719ecd780a52fb80a22b901ed5b6fe43fb45b
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/nlohmannjson
+	SECTION:=libs
+	CATEGORY:=Libraries
+	TITLE:=JSON for Modern C++
+	URL:=https://nlohmann.github.io/json/
+	BUILDONLY:=1
```
Comment (BKPepe) [inline libs/nlohmannjson/Makefile]: NACK. I agree with @karlp.

Comment (BKPepe) [inline libs/nlohmannjson/Makefile]: Two spaces instead of tab for indentation.

File: libs/ipfs-http-client/Makefile [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+	SECTION:=libs
+	CATEGORY:=Libraries
+	TITLE:=IPFS client library
+	URL:=https://github.com/vasild/cpp-ipfs-http-client
+	DEPENDS:= +libcurl +libstdcpp
+	PKG_BUILD_DEPENDS:=nlohmannjson
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]: Two spaces as indentation for each row here as well.

File: libs/ipfs-http-client/Makefile [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+PKG_SOURCE_DATE:=2019-11-05
+PKG_SOURCE_VERSION:=763e59ad698f3e3846f85df11e01c18ef3fbc401
+PKG_MIRROR_HASH:=a2b5721faf0d43ddb4a892245ef382666149c83f3f97e558e1a6acf2402fb9fd
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/$(PKG_NAME)/Default/description
+	IPFS (the InterPlanetary File System) is the Distributed Web. 
+	Specs, docs, sources, links: https://ipfs.io/ and https://github.com/ipfs.
+	This is Vasil Dimov's C++ IPFS HTTP API client library.
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]: Two spaces as indentation for each row in description.

File: libs/ipfs-http-client/Makefile [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+define Package/lib$(PKG_NAME)
+	SECTION:=libs
+	CATEGORY:=Libraries
+	TITLE:=IPFS client library
+	URL:=https://github.com/vasild/cpp-ipfs-http-client
+	DEPENDS:= +libcurl +libstdcpp
+	PKG_BUILD_DEPENDS:=nlohmannjson
+endef
+
+define Package/lib$(PKG_NAME)/description
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
define Package/libipfs-http-client/description
```

File: libs/ipfs-http-client/Makefile [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+	$(call Package/$(PKG_NAME)/Default/description)
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
	$(call Package/ipfs-http-client/Default/description)
```

File: libs/ipfs-http-client/Makefile [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+	This package contains shared library.
+endef
+
+define Package/$(PKG_NAME)-tests
+	SECTION:=utils
+	CATEGORY:=Utilities
+	TITLE:=IPFS client library tests
+	URL:=https://github.com/vasild/cpp-ipfs-http-client
+	DEPENDS:=+libipfs-http-client +libcurl
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]: Two spaces instead of tab.

File: libs/ipfs-http-client/Makefile [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+	DEPENDS:= +libcurl +libstdcpp
+	PKG_BUILD_DEPENDS:=nlohmannjson
+endef
+
+define Package/lib$(PKG_NAME)/description
+	$(call Package/$(PKG_NAME)/Default/description)
+	This package contains shared library.
+endef
+
+define Package/$(PKG_NAME)-tests
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
define Package/ipfs-http-client-tests
```

File: libs/ipfs-http-client/Makefile [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+	SECTION:=utils
+	CATEGORY:=Utilities
+	TITLE:=IPFS client library tests
+	URL:=https://github.com/vasild/cpp-ipfs-http-client
+	DEPENDS:=+libipfs-http-client +libcurl
+endef
+
+define Package/$(PKG_NAME)-tests/description
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
define Package/ipfs-http-client-tests/description
```

File: libs/ipfs-http-client/Makefile [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+	$(call Package/$(PKG_NAME)/Default/description)
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
	$(call Package/ipfs-http-client/Default/description)
```

File: libs/ipfs-http-client/Makefile [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+	This package contains library tests.
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
  This package contains library tests.
```

File: libs/ipfs-http-client/Makefile [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+endef
+
+define Package/$(PKG_NAME)-tests/install
+	$(INSTALL_DIR) $(1)/usr/bin
+	$(INSTALL_BIN) $(PKG_BUILD_DIR)/test/ipfs-* $(1)/usr/bin
+endef
+
+$(eval $(call BuildPackage,lib$(PKG_NAME)))
+$(eval $(call BuildPackage,$(PKG_NAME)-tests))
+
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]: The last empty row is not required. Remove it, please.

File: libs/ipfs-http-client/Makefile [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+endef
+
+define Build/InstallDev
+	$(INSTALL_DIR) $(1)/usr/include
+	$(INSTALL_DIR) $(1)/usr/lib
+	$(CP) $(PKG_BUILD_DIR)/include/* $(1)/usr/include
+	$(CP) $(PKG_BUILD_DIR)/libipfs-http-client.a $(1)/usr/lib
+endef
+
+define Package/lib$(PKG_NAME)/install
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
define Package/libipfs-http-client/install
```

File: libs/ipfs-http-client/Makefile [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+	$(INSTALL_DIR) $(1)/usr/lib
+	$(CP) $(PKG_BUILD_DIR)/libipfs-http-client.so* $(1)/usr/lib
+endef
+
+define Package/$(PKG_NAME)-tests/install
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
define Package/ipfs-http-client-tests/install
```

File: libs/ipfs-http-client/Makefile [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+	$(INSTALL_DIR) $(1)/usr/bin
+	$(INSTALL_BIN) $(PKG_BUILD_DIR)/test/ipfs-* $(1)/usr/bin
+endef
+
+$(eval $(call BuildPackage,lib$(PKG_NAME)))
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
$(eval $(call BuildPackage,libipfs-http-client))
```

File: libs/ipfs-http-client/Makefile [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+$(eval $(call BuildPackage,$(PKG_NAME)-tests))
```
Comment (BKPepe) [inline libs/ipfs-http-client/Makefile]:
```suggestion
$(eval $(call BuildPackage,ipfs-http-client-tests))
```

File: libs/nlohmannjson/Makefile (hunk 10 above)
Comment (BKPepe) [inline libs/nlohmannjson/Makefile]: Two spaces instead of tab.

Comment (BKPepe) [inline libs/nlohmannjson/Makefile]: No, remove it, please.

File: libs/nlohmannjson/Makefile (hunk 8 above)
Comment (BKPepe) [inline libs/nlohmannjson/Makefile]: You have here empty row as well, remove it please.

**Comment by LLE8** [comment]
About BUILDONLY.
Find strings "buildonly" in scripts/package-metadata.pl and scripts/metadata.pm. What do they mean?

**Comment by neheb** [comment]
Please squash these commits.

File: libs/ipfs-http-client/Makefile [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+  This is Vasil Dimov's C++ IPFS HTTP API client library.
+endef
+
+define Package/libipfs-http-client
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=IPFS client library
+  URL:=https://github.com/vasild/cpp-ipfs-http-client
+  DEPENDS:= +libcurl +libstdcpp
+  PKG_BUILD_DEPENDS:=nlohmannjson
```
Comment (neheb) [inline libs/ipfs-http-client/Makefile]: This does not belong here

Comment (LLE8) [reply libs/ipfs-http-client/Makefile]: Why?  I don't understand what you want to say.

Comment (neheb) [reply libs/ipfs-http-client/Makefile]: It belongs outside of the group.

Comment (LLE8) [reply libs/ipfs-http-client/Makefile]: Line PKG_BUILD_DEPENDS moved outside of the group

File: libs/ipfs-http-client/Makefile (hunk 5 above)
Comment (LLE8) [reply libs/ipfs-http-client/Makefile]: moved

File: libs/ipfs-http-client/Makefile [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+PKG_LICENSE_FILES:=LICENSE.MIT
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/vasild/cpp-ipfs-http-client.git
+PKG_SOURCE_DATE:=2019-11-05
+PKG_SOURCE_VERSION:=763e59ad698f3e3846f85df11e01c18ef3fbc401
+PKG_MIRROR_HASH:=a2b5721faf0d43ddb4a892245ef382666149c83f3f97e558e1a6acf2402fb9fd
+PKG_BUILD_DEPENDS:=nlohmannjson
+
+include $(INCLUDE_DIR)/package.mk
```
[the remark this answers is not in the document]
Comment (LLE8) [reply libs/ipfs-http-client/Makefile]: CMake is overkill for program of two C++ files
https://github.com/vasild/cpp-ipfs-http-client/issues/27

**Comment by LLE8** [comment]
I agree, BUILDONLY works as intended, but it's undocumented. Ask @karlp and @BKPepe before i will give it back.

File: libs/nlohmannjson/Makefile [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+PKG_HASH:=e109cd4a9d1d463a62f0a81d7c6719ecd780a52fb80a22b901ed5b6fe43fb45b
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/nlohmannjson
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=JSON for Modern C++
+  URL:=https://nlohmann.github.io/json/
+endef
```
[the remark this answers is not in the document]
Comment (LLE8) [reply libs/nlohmannjson/Makefile]: I agree, BUILDONLY works as intended, but it's undocumented. 
Find strings "buildonly" in scripts/package-metadata.pl and scripts/metadata.pm. What do they mean? Ask @karlp and @BKPepe before i will give it back.

File: libs/ipfs-http-client/Makefile (hunk 26 above)
[the remark this answers is not in the document]
Comment (neheb) [reply libs/ipfs-http-client/Makefile]: OTOH CMake will simplify the Makefile by getting rid of the InstallDev section (replaced by CMAKE_INSTALL).

[the remark this answers is not in the document]
Comment (LLE8) [reply libs/ipfs-http-client/Makefile]: This section contains 4 lines, there is nothing to simplify.

**Pull request merged**